Built upon base code

BenSt-Clair opened the pull request:
I've created some new classes including a combat system. Please let me know what you think! :)

magicalpapaya58717 committed:
I am now pushing these changes to the development branch on github

notxcharles reviewed (commented):
Overall 
This is a really, really good project. You have obviously spent a lot of time and effort on it. There's good proof of encapsulation and abstractions. The classes represent just one type of object, too. There are base classes that demonstrate the use of class inheritance within your project. The random messages that are shown makes every playthrough of the game unique and enjoyable

Additional Feedback:
I'd like to see each class have it's own separate file, I think this would help with readability and finding the locations of some classes more quickly
From a user's point of view, it would be nice and more convenient if I were told a list of accepted input when asked a question. 
Not all functions and variables had a consistent naming and casing. I'd recommend that each function uses PascalCase and variables camelCase
Some classes felt like they could benefit from constructor overloading


Requirements:
1. Classes and Objects:
- [x] Game.cs file handles the game flow and initialises the player and one simple room
	- No file explicitly named game.cs but there was a file with a different name that achieved the same functionality
- [x] Player.cs tracks the player's name and a signle attribute, such as health or a basic inventory
	- Player has a name, health value and inventory. Name and Health have a getter and a setter (private setter for Name).
- [x] Room.cs represents a single room in the game with a description and possibly an item
	- Many items can be in the room, room has a name and description

2. Encapsulation and Abstraction
- [x] Use private fields and provide get/set methods where necessary
	- Used within most classes where needed

3. Methods & Constructors
- [x] Create constructors for the Player and Room classes to initialise their attributes.
	- All classes
	Implement basic methods:
	- [x] Player.PickUpItem(): Adds a single item to the player's inventory.
		- Ability to pick up an item, manage player's inventory and choose an item from the inventory
	- [x] Room.GetDescription(): Returns the description of
		- Room.GetDescription() returns a description of the room

4. Basic Game Interaction
- [x] The player starts in a single room and can:
	- [x] View the room's description.
		- User is given a description of the room at the start of the round
	- [x] Display their current status (e.g., inventory and health).
		- Player.cs is shown their health and can request a list of items in their inventory

5. Error Checking:
- [x] Add simple error checking, such as preventing the player from entering a room that does not exist.
	- Multiple examples of error checking exist within each class

gab-rielb reviewed (commented):
### Overall Thoughts
Overall, you have clearly thought out all aspects of your game, going above and beyond in some cases. This passion has definitely shown through all the different intricacies throughout your code. However, there are some trade-offs which comes with a high level of complexity of code which I would like to point out.

### Criticisms

- **Main Program** class is doing too much in my opinion. The entire game logic is contained in this class making it quite difficult to follow. Consider breaking the code down into seperate functions. For example, a function for initialising a game, creating a player, generating stats etc. This will not only improve readability but would be quite sustainable for expanding the game in the future.

- **Weapon.Attack** is an example of a method that contains many nested if statements, making the code quite convuluted. I'd suggest again potentially splitting up the code in to functions (e.g CalculateDamage) or consider using switch statements.

- **Player Class** could do with moving player-specific logic from the main program into it (e.g player creation, inventory management).

- **The complexity** deviates from the brief as it emphasises basic Player and Room classes. The current code has complex functionalities within these classes, like intricate descriptions and various item interactions, which deviate from the "basic" requirement. Which isn't a bad thing, however, I would personally focus on ensuring a well-structured and maintained "basic" code as a foundation to then build on further complexities. As with such detailed code it would be very easy to overlook certain basic features. For clarity, I am not saying you are overlooking these features, however, I am pointing out the increased risks and other potential trade-offs from having such complex code. Furthermore, having an extremely solid foundation will most definetly benefit you for assessment 2 where you will be required to go into even more detail.

- **Comments** are included in your program but they could be more descriptive in explaining "why" rather than just "what" the code is doing. But this is only a small nuance.

- **Game.cs** is missing. I know most of the functionality is coded within the main program however, potentially using a "Game" class for the game logic could increase maintainability as well as readability. Also Christos did mention that it was his preference for people to "add/modify" rather than "remove" parts of the code. I assume this won't be an issue as your program still carries out the parts of the brief as required. I would personally just be cautious.

### Praise

- **Combat** is very creative and definetly engaging, some very good ideas here.

- **Descriptions** are very creative and add to the immersion of the game for sure.

- **Customisations** of characters is a good idea which adds a sense of personality to the game.

- **Encapsulation** is well implemented. Good use of get-set functions as well.

- **Object Oriented** approach is taken quite well.



_p.s I tried to run your code on my computer and it came up with a few errors, so i'd suggest testing your code on multiple computers before submission :)_

BenSt-Clair committed:
Add files via upload

Added <GenerateAssemblyInfo>false line of code to the project file.
Now AssemblyInfo shouldn't be automatically generated and the corresponding duplicate build errors shouldn't occur when downloaded.
